Add python indentation script that external projects can use.

vishalkenchan opened the pull request:
I found it difficult to understand bash script to indent/format code. So, I wrote a similar script in python that uses multiple threads to format code using clang-format. I thought may be someone might be interested in using this script in their own projects.

Note that some of the functionalities of `indent-all.sh` script are missing in this python script (such as fixing `.inst.in` files, fixing permissions, or conversion of DOS formatted files to unix file format etc).

Not sure if this script should be in deal.II. If it should belong in deal.II should it be in `cmake/scripts` directory?

davydden commented:
> Not sure if this script should be in deal.II. If it should belong in deal.II should it be in cmake/scripts directory?

I don't know what happens with files in `contrib/utilities/*`, but we might need to install it manually as we want to let users use it.

davydden commented on contrib/utilities/indent.py:
maybe to add a git command that can set this as a pre-commit hook? Just save users' time from looking.

davydden commented on contrib/utilities/indent.py:
```diff
@@ -0,0 +1,260 @@
+
```
please copy deal.II copyright header (last time I added `.py` I copied the header verbatim and added comment character in editor).

davydden commented on contrib/utilities/indent.py:
maybe something like:

Run clang-format on all file in a list of directories that satisfy the regex.

davydden commented on contrib/utilities/indent.py:
do we want to say that it's w.r.t. pwd? 

Comma-delimited list of directories  with respect to the current/present working directory

davydden commented:
@vishalkenchan we should install this script. See https://github.com/dealii/dealii/blob/master/cmake/scripts/CMakeLists.txt#L48 . This will end up in `prefix/share/deal.II/scripts`.

davydden commented:
Also please add a changelog.

drwells commented:
Thanks for doing this! I have a few minor stylistic comments.

I ran pylint on the script and got the following result:
```
[drwells@learch utilities]$ pylint indent.py 
************* Module indent
indent.py:39:20: C0326: No space allowed before bracket
def parse_arguments ():
                    ^ (bad-whitespace)
indent.py:40:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
[...]
indent.py:76:69: C0326: Exactly one space required after comma
def check_clang_format_version(args, compatible_version_list=['6.0.0','6.0.1']):
                                                                     ^ (bad-whitespace)
indent.py:77:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
[...]
indent.py:86:0: C0330: Wrong continued indentation (add 2 spaces).
                                                    '--version'])
                                                    ^ | (bad-continuation)
indent.py:87:0: W0311: Bad indentation. Found 6 spaces, expected 12 (bad-indentation)
indent.py:88:0: C0330: Wrong continued indentation (add 1 space).
                                clang_format_version,
                                ^| (bad-continuation)
indent.py:89:0: C0330: Wrong continued indentation (add 1 space).
                                re.IGNORECASE).group(1)
                                ^| (bad-continuation)
indent.py:90:0: W0311: Bad indentation. Found 6 spaces, expected 12 (bad-indentation)
indent.py:91:0: W0311: Bad indentation. Found 8 spaces, expected 16 (bad-indentation)
indent.py:92:0: C0330: Wrong hanging indentation (add 2 spaces).
          """
          ***
          ***   No compatible clang-format program found.
          ***
          ***   Install any of the following versions
          ***""" ' '.join(compatible_version_list)
          ^ | (bad-continuation)
indent.py:98:0: C0330: Wrong hanging indentation.
          )
        | ^ | (bad-continuation)
indent.py:99:0: W0311: Bad indentation. Found 4 spaces, expected 8 (bad-indentation)
[...]
indent.py:103:0: C0330: Wrong hanging indentation (add 2 spaces).
      """
      ***
      ***   No clang-format program found.
      ***
      ***   You can run
      ***       'contrib/utilities/download_clang_format'
      ***   or
      ***       'contrib/utilities/compile_clang_format'
      ***   to install a compatible binary into 'contrib/utilities/programs'.
      ***
      """
      ^ | (bad-continuation)
indent.py:114:0: C0330: Wrong hanging indentation.
      )
    | ^ | (bad-continuation)
indent.py:117:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
[...]
indent.py:142:32: C0326: No space allowed before bracket
    _, temp_file_name = mkstemp (dir=temp_dir,
                                ^ (bad-whitespace)
indent.py:146:0: W0311: Bad indentation. Found 4 spaces, expected 8 (bad-indentation)
indent.py:146:20: C0326: No space allowed before bracket
    shutil.copyfile (full_file_name, temp_file_name)
                    ^ (bad-whitespace)
indent.py:151:0: W0311: Bad indentation. Found 4 spaces, expected 8 (bad-indentation)
[...]
indent.py:154:22: C0326: No space allowed before bracket
      subprocess.call (apply_clang_format_str, shell=True)
                      ^ (bad-whitespace)
indent.py:155:0: W0311: Bad indentation. Found 4 spaces, expected 8 (bad-indentation)
[...]
indent.py:169:14: C0326: No space allowed before bracket
        print (full_file_name, " - file indented incorrectly")
              ^ (bad-whitespace)
indent.py:170:0: W0311: Bad indentation. Found 8 spaces, expected 16 (bad-indentation)
indent.py:170:18: C0326: No space allowed before bracket
        os.remove (temp_file_name)
                  ^ (bad-whitespace)
indent.py:171:0: W0311: Bad indentation. Found 6 spaces, expected 12 (bad-indentation)
indent.py:172:0: W0311: Bad indentation. Found 8 spaces, expected 16 (bad-indentation)
indent.py:172:20: C0326: No space allowed before bracket
        shutil.move (temp_file_name, full_file_name)
                    ^ (bad-whitespace)
indent.py:178:0: W0311: Bad indentation. Found 4 spaces, expected 8 (bad-indentation)
indent.py:180:12: C0326: No space allowed before bracket
def process (args):
            ^ (bad-whitespace)
indent.py:181:0: W0311: Bad indentation. Found 2 spaces, expected 4 (bad-indentation)
[...]
indent.py:208:17: C0326: No space allowed before bracket
  for _ in range (n_threads):
                 ^ (bad-whitespace)
indent.py:209:0: W0311: Bad indentation. Found 4 spaces, expected 8 (bad-indentation)
indent.py:210:0: C0330: Wrong continued indentation (remove 1 space).
                                      args=(args, task_queue, tmpdir))
                                     |^ (bad-continuation)
indent.py:211:0: W0311: Bad indentation. Found 4 spaces, expected 8 (bad-indentation)
[...]
indent.py:1:0: C0111: Missing module docstring (missing-docstring)
indent.py:13:0: E0401: Unable to import 'Queue' (import-error)
indent.py:76:0: W0102: Dangerous default value [] as argument (dangerous-default-value)
indent.py:76:31: W0621: Redefining name 'args' from outer scope (line 238) (redefined-outer-name)
indent.py:99:4: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
indent.py:116:16: W0621: Redefining name 'args' from outer scope (line 238) (redefined-outer-name)
indent.py:158:4: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
indent.py:180:13: W0621: Redefining name 'args' from outer scope (line 238) (redefined-outer-name)
indent.py:197:2: W1201: Specify string format arguments as logging function parameters (logging-not-lazy
)                                                                                                      
indent.py:237:2: C0103: Constant name "start" doesn't conform to UPPER_CASE naming style (invalid-name)
indent.py:238:2: C0103: Constant name "args" doesn't conform to UPPER_CASE naming style (invalid-name)
indent.py:258:2: C0103: Constant name "finish" doesn't conform to UPPER_CASE naming style (invalid-name)
indent.py:260:2: W1201: Specify string format arguments as logging function parameters (logging-not-lazy
)                                                                                                      

------------------------------------
```

Furthermore, would you please move the comment at the top into a top-level docstring and also add a `#!/usr/bin/env python` to the first line?

davydden commented:
p.s. @vishalkenchan If you install `python` package in VSCode by Microsoft, it should help with `pylint` and will automatically show you the problems.

drwells commented:
Some of the pylint remarks are rather pedantic (like that `finish` should be in all caps) but we really should fix at least the whitespace issues (the rest of our python code is indented in a style that makes pylint happy).

vishalkenchan commented:
Thank you @drwells for your suggestion! pylint seems to be happy now. I wonder if I should move `indent.py` to `cmake/scripts` but actually one could use this script while developing deal.II as well.

davydden commented:
> I wonder if I should move indent.py to cmake/scripts but actually one could use this script while developing deal.II as well.

i would keep them where they are because indeed we one can use it instead of the bash script.

masterleinad commented:
> i would keep them where they are because indeed we one can use it instead of the bash script.

But I see no reason for using it for `deal.II`. The bash script can do more things, doesn't need an additional tool and you can invoke it as a build target.

davydden commented:
@drwells ping.

drwells commented:
There are a few more (small) issues with this script:
1. I don't have the `Queue` module on my machine: I think this was renamed in python3:
https://stackoverflow.com/questions/29687837/queue-importerror-in-python-3
2. There is still an extra space in line 210.
3. There are some of the dreaded unicode/bytes conversion problems when running with python3:
```
[drwells@learch build]$ python ../contrib/utilities/indent.py
Traceback (most recent call last):
  File "../contrib/utilities/indent.py", line 300, in <module>
    ['6.0.0', '6.0.1'])
  File "../contrib/utilities/indent.py", line 131, in check_clang_format_version
    re.IGNORECASE).group(1)
  File "/usr/lib/python3.7/re.py", line 183, in search
    return _compile(pattern, flags).search(string)
TypeError: cannot use a string pattern on a bytes-like object
```

There may be more issues but I cannot get past that with python3.

vishalkenchan commented:
> 1. I don't have the Queue module on my machine: I think this was renamed in python3:
https://stackoverflow.com/questions/29687837/queue-importerror-in-python-3

Fixed it as suggested.
> 2. There is still an extra space in line 210.

Fixed.
> 3. There are some of the dreaded unicode/bytes conversion problems when running with python3.

It seems that python3's `subprocess.check_output()` yields bytes-like output. I think using a `decode()` should have fixed it.

@drwells I now tested the script with both python2 and python3.

drwells commented:
Okay, I can now run the script and it exits without an unintended exception. I now get:

```
[drwells@learch build]$ python ../contrib/utilities/indent.py
ERROR:root:6.0.0
                    ***
                    ***   No compatible clang-format program found.
                    ***
                    ***   Install any of the following versions
                    *** 6.0.1
NoneType: None
INFO:root:Number of threads picked up: 3
INFO:root:Finished code formatting in: 1.353010 seconds.
```

I am not sure where `NoneType: None` came from, but the call to `join` on line 139 is wrong (it intercalates the valid version numbers with the errors; I think you meant to write `",".join(compatible_version_list)`).

I think the program should abort if the versions don't work for compatibility with the shell script (I don't know what it is doing for the last 1.35 seconds of runtime; its not indenting anything).

Other than that, I can confirm that it works as advertised.

vishalkenchan commented:
Oh yes. I thought `logging.exception()` causes `SystemExit()`. Now I just use `sys.exit()`.

davydden commented:
@vishalkenchan could you please squash into a single commit?

vishalkenchan commented:
@davydden Done.

davydden commented:
Let’s wait for @drwells to press the green button if all is good on.

drwells commented:
Sorry, I should have been more specific with the `join` comment: we now get 
```
6.0.0
                    ***
                    ***   No compatible clang-format program found.
                    ***
                    ***   Install any of the following versions
                    ***,6.0.1
```
the correct code is
```python
                sys.exit(
                    """
                    ***
                    ***   No compatible clang-format program found.
                    ***
                    ***   Install any of the following versions:
                    ***   """ + ", ".join(compatible_version_list)
                )
```

tjhei commented on doc/news/changes/minor/20191301VishalBoddu:
```diff
@@ -0,0 +1,3 @@
+New: Indentation script (written in python) to format code in external projects that use deal.II
```
can you add the relative path+name of this script in here?

davydden commented:
@tjhei can we merge this?

tjhei commented:
:+1:
